### PR TITLE
Adds new version of Issue Labeler

### PR DESCRIPTION
**Pull request by maryamariyan**
cc: @Anipik

### CODE REVIEW

File: src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            var lines = File.ReadAllLines(inputPath);
+            int totalCount = lines.Length;
+
+            // have at least 1000 elements
+            Debug.Assert(totalCount > 1000);
+            int numInTrain = (int)(lines.Length * 0.8);
+            int numInValidate = (int)(lines.Length * 0.1);
+
+            // 80% into train dataset
+            SaveFromXToY(
```
Comment (Anipik) [inline src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs]: have you tried using ML.net here ? they might have some useful apis to do this

File: src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            lines = span.ToArray();
+            File.AppendAllLines(output, lines.Where(x => int.TryParse(x.Split('\t')[3], out int isPrAsNumber) && isPrAsNumber == 1).ToArray());
+        }
+
+        /// <summary>
+        /// saves to file a subset containing only issues
+        /// </summary>
+        /// <param name="input">path to the reference dataset</param>
+        /// <param name="output">the output to store the new dataset</param>
+        public void OnlyIssues(string input, string output)
```
Comment (Anipik) [inline src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs]: We can combine onlyPrs and OnlyIssues using a boolean flag

File: src/Microsoft.DotNet.Github.IssueLabeler/Models/Predictor.cs [hunk 3]
```diff
@@ -10,22 +10,40 @@ namespace Microsoft.DotNet.GitHub.IssueLabeler
...
                 MLContext mlContext = new MLContext();
-                ITransformer mlModel = mlContext.Model.Load(ModelPath, out DataViewSchema inputSchema);
-                predEngine = mlContext.Model.CreatePredictionEngine<GitHubIssue, GitHubIssuePrediction>(mlModel);
+                ITransformer mlModel = mlContext.Model.Load(IssueModelPath, out DataViewSchema inputSchema);
+                issuePredEngine = mlContext.Model.CreatePredictionEngine<IssueModel, GitHubIssuePrediction>(mlModel);
             }
 
-            GitHubIssuePrediction prediction = predEngine.Predict(issue);
+            GitHubIssuePrediction prediction = issuePredEngine.Predict(issue);
             float[] probabilities = prediction.Score;
```
Comment (Anipik) [inline src/Microsoft.DotNet.Github.IssueLabeler/Models/Predictor.cs]: wecab move this duplicated code to a separate function

File: src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DiffHelper.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        /// folder names taken from parsing the fileDiffs, while keeping track of the number of times such a folder name was repeated
+        /// </summary>
+        public Dictionary<string, int> FolderNames { get; } = new Dictionary<string, int>();
+
+        private void SetupFolderParts()
+        {
+            FolderNames.Clear();
+            Folders.Clear();
+            string folderWithDiff, subfolder;
+            string[] folderNames;
```
Comment (Anipik) [inline src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DiffHelper.cs]: can we have some different name to differentiate between the dictionary and array

File: src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs (hunk 1 above)
Comment (maryamariyan) [reply src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs]: pushed a commit using DataFrame instead

File: src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            int bodyIndex = dataFrame.Columns.IndexOf("Description");
+            int numMentionsIndex = dataFrame.Columns.IndexOf("NumMentions");
+            int userMentionsIndex = dataFrame.Columns.IndexOf("UserMentions");
+            int filePathsIndex = dataFrame.Columns.IndexOf("FilePaths");
+            int fileCountIndex = dataFrame.Columns.IndexOf("FileCount");
+            int filesIndex = dataFrame.Columns.IndexOf("Files");
+            int filenamesIndex = dataFrame.Columns.IndexOf("Filenames");
+            int fileExtensionsIndex = dataFrame.Columns.IndexOf("FileExtensions");
+            int folderNamesIndex = dataFrame.Columns.IndexOf("FolderNames");
+            int foldersIndex = dataFrame.Columns.IndexOf("Folders");
```
Comment (pgovind) [inline src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs]: Nothing wrong with this, but if you want to reduce code later, you can refer to a column by it's name directly

File: src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        }
+
+        /// <summary>
+        /// flattens a dictionary to be repeated in a space separated format
+        /// </summary>
+        /// <param name="textToFlatten">a dictionary containing text and number of times they were repeated</param>
+        /// <returns>space delimited text</returns>
+        public string FlattenIntoColumn(Dictionary<string, int> folder)
+        {
+            _folderSb.Clear();
```
Comment (pgovind) [inline src/Microsoft.DotNet.Github.IssueLabeler/Helpers/DatasetHelper.cs]: You don't have to do it right now, but this _might_ be replaceable with a regex? Also, who is calling this? All the `FlatternIntoColumn` calls seem to pass in a string?

**Comment by ericstj** [comment]
@maryamariyan Is this one ready to go in?

**Comment by maryamariyan** [comment]
> @maryamariyan Is this one ready to go in?

yes. I'll merge as it has approval as well once CI gets completed.

**Pull request merged**